doc/neps/nep-0043-extensible-ufuncs.rst

```diff
@@ -0,0 +1,631 @@
+2. A concrete Unit DType which parametrizes the numeric dtype making it
+   ``Unit(float64, "m")``.
+
+The difference between the two, is that in the first case a ``UFuncImpl``
+can be created which *wraps* a single existing ``UFuncImpl`` and after that
+dispatching can be cached.  The creation of this ``UFuncImpl`` can further
+occur dynamically in the *promotion* step.
+To fall back to the existing loop, we allow creation of a wrapped loop
+from an existing (bound) ``UFuncImpl``.
+
+This wrapped loop will have two additional methods:
+
+* ``view_inputs(Tuple[DType]: input_descr) -> Tuple[DType]`` replacing the
+  user input descriptors with descriptors matching the wrapped loop.
+  It must be possible to *view* the inputs as the output.
+* ``wrap_outputs(Tuple[DType]: input_descr) -> Tuple[DType]`` replacing the
+  resolved descriptors with with the desired actual loop descriptors.
+  The original ``resolve_descriptors`` function will be called between these
+  two calls, so that the output descriptors may not be set in the first call.
+
+The ``view_inputs`` method allows passing the correct inputs into the
+original ``resolve_descriptors`` function, while ``wrap_outputs`` ensures
+the correct descriptors are used for output allocation and input buffering casts.
+
+The second use-case can be implemented within the inner-loop and
+``resolve_descriptors`` functions.
+This will incur additional overheads, partially because the loop dispatch
+effectively has to be done twice, but mostly because an efficient
+implementation will require direct access to the original
+setup/teardown, etc. functions.
+Such functionality can be added at a later time.
+
+
+Details for ``resolve_descriptors``
+"""""""""""""""""""""""""""""""""""
+
+*TODO: Should this function also get the full set of information which
+I want to pass in to the setup/teardown/inner-loop?  On the one hand, much
+of the information is not yet available/defined (or is set here).  On the
+other hand, some of the info is useful, and it may be nice to just have
+a homogeneous calling convention.*
+
+The UFunc machinery must know the correct dtypes to use before arrays can
+be allocated. The arrays creation itself is desirable to happen before any
+setup functionality to allow potential choosing of an optimized loops.
+
+**Notes:**
+
+There are a few possible extension to this function.  Currently, it also
+takes care of casting in general.  This is not necessary, however, it
+would be possible to extend the signature with casting indication for
+*outputs*.
+This would allow registering e.g. ``float64 + float64 -> float32`` as an explicit
+(faster) loop while indicating that it is an unsafe cast on the result array,
+which requires the user to specify ``casting="unsafe"``.
+
+The current design allows such a specialized loop (with access to the
+initially private ``setup``), from within the ``float64+float64->float64``
+implementation only.
+
+
+Bound UFuncImpl
```

Comment:
This section ended up confusing me no end. It is not obvious to me that it is like a bound method where the ufunc and dtypes combine to some form of `self`, unless one thinks of each `UFuncImpl` as some type of class that has `__init__(self, ufunc, dtypes)`, so that all the methods can just refer to `self.ufunc` and `self.dtypes`.

Indeed, I'm sufficiently confused that it makes me wonder whether it might not be helpful to have a mock-up of a python implementation of a `UFuncImpl` and on how one would use it for an existing function (the units case would be a good one), as well as how one would write a new `UFunc` taking perhaps an existing `UFuncImpl` (say, a `sincos` implementation). (This echoes a point made by @mattip in the original thread...)

Reply:
Yeah, I guess you could see it that way with an `__init__(self, ufunc, dtypes)`, although, I do not think such an init has to technically be exposed to the user.  Anyway, I have to think about trying to clear this up, I was not sure I like the mental model around having an `__init__`, it somewhat implies there is a full blown (subclass-able?) class, something that I would like to avoid at least for now.

Reply:
Agreed that the way it is exposed to a python user does not have to be decided now. My main comment here was really that I found the present text confusing (unlike the rest). Might it make sense to move this piece to a separate very preliminary discussion of how things might be exposed in python?

 Or, not as good but perhaps more practical, remove it and just have a note up front that a python implementation is not yet fully considered but that it is attempted to make sure nothing would stop it?

Reply:
I have expanded/refactored this section, and more importantly tried to add a super-highlevel-get-the-feel for things section at the start, which happens to include `resolve_impl` (I don't know how much that helps, and I guess you are the wrong person to ask by now).

In any case, this section is restructured, and I tried to point out that it is a bit nitty-gritty for the interested reader (I am unsure about it, but it helped me be clear about who has what information)